File: dynamic_listings/C++/default/computer_science/inheritance_example.cpp
class Shape{
    // An abstract shape class
    // An abstract function that will be overridden by subclasses
    virtual float area() = 0;
    // An abstract function that will be overridden by subclasses
    virtual float perimeter() = 0;
};

class Rectangle : public Shape{
    // A simple rectangle class
    float width;
    float height;

    Rectangle(float w, float h){
        width = w;
        height = h;
    }

    float area(){
        // Returns the Area of the rectangle
        return width * height;
    }

    float perimeter(){
        // Returns the Perimeter of the rectangle
        return 2 * (width + height);
    }
};

class Circle : public Shape{
    // A simple circle class
    float radius;

    Circle(float r){
        radius = r;
    }

    float area(){
        // Returns the Area of the circle
        return 3.1415 * 3.1415 * radius;
    }

    float perimeter(){
        // Returns the circumference of the circle
        return 2 * 3.1415 * radius;
    }
};

File: dynamic_listings/python/default/computer_science/inheritance_example.py
from math import pi


class Shape:
    # An abstract shape class
    def area() -> float:
        # An abstract function that will be overridden by subclasses
        raise NotImplementedError

    def perimeter() -> float:
        # An abstract function that will be overridden by subclasses
        raise NotImplementedError


class Rectangle(Shape):
    # A simple rectangle class
    width: float = 0.0
    height: float = 0.0

    def __init__(self, width: float, height: float) -> None:
        self.width = width
        self.height = height

    def area(self) -> float:
        # Returns the Area of the rectangle
        return self.width * self.height

    def perimeter(self) -> float:
        # Returns the Perimeter of the rectangle
        return 2 * (self.width + self.height)


class Circle(Shape):
    # A simple circle class
    radius: float = 0.0

    def __init__(self, radius: float) -> None:
        self.radius = radius

    def area(self) -> float:
        # Returns the Area of the circle
        return pi * self.radius ** 2

    def perimeter(self) -> float:
        # Returns the circumference of the circle
        return 2 * pi * self.radius
